### PR TITLE
Updated Changefeed to be able to read last consumable. Also unified cursor.

### PR DESCRIPTION
**Pull request by gapra-msft**
Cursor format:

{
  "CursorVersion": 1,
  "UrlHost": "azstoragesdkaccount.blob.core.windows.net",
  "EndTime": "2020-07-31T00:00:00Z",
  "CurrentSegmentCursor": {
    "ShardCursors": [
      {
        "CurrentChunkPath": "log/00/2020/07/30/2300/00000.avro",
        "BlockOffset": 96253,
        "EventIndex": 0
      }
    ],
    "CurrentShardPath": "log/00/2020/07/30/2300/",
    "SegmentPath": "idx/segments/2020/07/30/2300/meta.json"
  }
}

### CODE REVIEW

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java [hunk 1]
```diff
@@ -100,25 +116,24 @@
...
-                    }
-                    return Mono.just(this.lastConsumable);
-                } catch (IOException e) {
-                    return FluxUtil.monoError(logger, new UncheckedIOException(e));
+            .flatMap(jsonNode -> {
+                /* Last consumable time. The latest time the changefeed can safely be read from.*/
+                OffsetDateTime lastConsumableTime = OffsetDateTime.parse(jsonNode.get("lastConsumable").asText());
+                /* Soonest time between lastConsumable and endTime. */
+                OffsetDateTime safeEndTime = this.endTime;
+                if (lastConsumableTime.isBefore(endTime)) {
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java:126]: endTime is/should be optional parameter. How do we handle that case? I.e. we should default to lastConsumable if end time is null . Also if end time is null we shouldn't populate it in the cursor.

File: sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroObject.java [hunk 2]
```diff
@@ -35,8 +42,8 @@ public long getBlockOffset() {
     /**
      * @return The index of the object in the block.
      */
-    public long getObjectBlockIndex() {
-        return objectBlockIndex;
+    public long getEventIndex() {
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroObject.java:45]: I'd go back to `ObjectIndex`. Avro doesn't know it's used in CF context. It's supposed to be general purpose parser.

File: sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroParser.java [hunk 3]
```diff
@@ -106,31 +103,26 @@ private void onFilteredHeader(Object header) {
...
-        AvroSchema.checkType("objectBlockIndex", index, Long.class);
-        long threshold = (Long) index;
+        AvroSchema.checkType("threshold", threshold, Long.class);
 
-        this.blockOffset = this.state.getSourceOffset();
-        this.objectBlockIndex = 0;
-        AvroBlockSchema blockSchema = new AvroBlockSchema(
+        final AvroBlockSchema blockSchema = new AvroBlockSchema(
             this.objectType,
+            (Long) threshold,
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroParser.java:122]: maybe this should be called beginIndex or beginObjectIndex. Threshold sounds more like upper boundary.

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java:126]: I think ChangefeedFactory deals with this - we default to the Max time

OffsetDateTime start = startTime == null ? OffsetDateTime.MIN : startTime;
        OffsetDateTime end = endTime == null ? OffsetDateTime.MAX : endTime;

        return new Changefeed(this.client, start, end, null, segmentFactory);

Comment (gapra-msft) [reply sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java:126]: I feel like if endTime is null, the user just wants to go forever so I don't see the issue in doing this

Comment (kasobol-msft) [reply sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java:126]: great. sounds good.

File: sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroObject.java (hunk 2 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroObject.java]: done

File: sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroParser.java (hunk 3 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-internal-avro/src/main/java/com/azure/storage/internal/avro/implementation/AvroParser.java]: done

**Pull request merged**